src/core/datastore/sync/sync-manager.js

```diff
@@ -60,8 +62,31 @@ export class SyncManager {
     if (isEmpty(collection)) {
       return Promise.reject(new KinveyError('Invalid or missing collection name'));
     }
-    return this._fetchItemsFromServer(collection, query, options)
-      .then(entities => this._replaceOfflineEntities(collection, query, entities));
+
+    return this.getSyncItemCountByEntityQuery(collection, query)
+      .then((count) => {
+        if (count > 0) {
+          // TODO: I think this should happen, but keeping current behaviour
+          // const msg = `There are ${count} entities awaiting push. Please push before you attempt to pull`;
+          // return Promise.reject(new KinveyError(msg));
+          return this.push(collection, query)
+            .then(() => this.getSyncItemCountByEntityQuery(collection, query));
+        }
+
+        return count;
+      })
+      .then((count) => {
+        // Throw an error if there are still items that need to be synced
+        if (count > 0) {
+          return Promise.reject(
+            new SyncError('Unable to fetch the entities on the backend.'
+              + ` There are ${count} entities that need to be synced.`)
+          );
+        }
+
+        return this._fetchItemsFromServer(collection, query, options)
+          .then(entities => this._replaceOfflineEntities(collection, query, entities));
```

Comment:
After seeing the changed `_fetchItemsFromServer()`, `_replaceOfflineEntities()` has to change too - it doesn't need to delete anything. If `_fetchItemsFromServer()` worked as expected, it should have deleted everything that needs deleting. Currently, it has the wrong filter, so this delete inside `_replaceOfflineEntities()` is the one that is doing the actual work.